### PR TITLE
allow using multiprocessing to parallelize pycbc_inspiral over templates

### PR DESCRIPTION
**Pull request by ahnitz**

### CODE REVIEW

**Comment by ahnitz** [comment]
The point of this is that in some cases it is nicer to use multiple cores by simply using multiprocessing instead of multiple threads. The throughput is not increased, but there is some additional flexibility in resource allocation on clusters (i.e. can use more memory per job, can run few jobs, etc).

**Comment by ahnitz** [comment]
Will remove WIP label once I've had a chance to test the output more carefully for correctness.

**Comment by ahnitz** [comment]
@spxiwh @josh-willis I think this is ready for review now. There are a couple minor changes I should flag and one more major change that we should discuss. 

First, some minor things

1) I've removed the cluster-method option. It now only handles the case of providing a window. The old cluster over template was not really being handled consistently and was making the code a bit messier. If there is a strong preference I could add it back in (but it would need to also be fixed as one other part of the code was implicitly assuming you chose window). My preference is simply to take it out and let 'window' be the choice now until / if something else is needed. 

2) I've rearranged the code to run in templates in blocks (so it can be parallelized), by default this will basically just do the same order as normal, however, it does mean that interactions with the option for condensing the triggers now need to be done carefully. I think @josh-willis added those. If you choose the trigger handling option to be a multiple then it should be OK, or if you use the default case of single-core then it also won't matter.

3) The performance metrics don't take into account this core useage. Should it? I think it wouldn't be hard to add in. What's your preference @josh-willis 

Finally, what I would consider a more major issue. 

1) 'sigmasq' was not be written to the output files correctly! I think this is an old bug (or perhaps just forgotten by me bad choice). It essential saved the value of sigmasq that each template had during the first segment *only*. This probably wasn't noticed as jobs are usually short enough that where we use this information downstream may not have been a big problem, but it's probably sub-optimal. Note, that this did not affect the sigmasq used to actually calculate SNRs etc, that was done on a segment/psd basis, it's just the value of sigmasq associated with a trigger and written to disk that has a problem. 

In this PR (so far) I have *only* fixed the code in pycbc inspiral to save sigmasq as used to calculate each trigger. This issue appears to also probably exist in other codes that use the eventmanager class as evident by the logic there. So the precessing inspiral / skymax stuff probably also has this feature. I haven't checked everywhere it might be used. @spxiwh What do you think?

File: bin/pycbc_inspiral [hunk 1]
```diff
@@ -197,6 +196,12 @@ parser.add_argument("--checkpoint-exit-maxtime", type=int,
                          " time is exceeded. Default is no checkpointing.")
 parser.add_argument("--checkpoint-exit-code", type=int, default=77,
                     help="Exit code returned if exiting after a checkpoint")
+parser.add_argument("--multiprocessing", type=int,
+                    help="Parallelize over multiple processes, note this is"
+                         "separate from threading using the proc. scheme")
+parser.add_argument("--template-groupby", type=int, default=1,
+                    help="Number of template to group together for parallel"
```
Comment (spxiwh) [inline bin/pycbc_inspiral:203]: templates, typo

File: bin/pycbc_inspiral [hunk 2]
```diff
@@ -197,6 +196,12 @@ parser.add_argument("--checkpoint-exit-maxtime", type=int,
                          " time is exceeded. Default is no checkpointing.")
 parser.add_argument("--checkpoint-exit-code", type=int, default=77,
                     help="Exit code returned if exiting after a checkpoint")
+parser.add_argument("--multiprocessing", type=int,
+                    help="Parallelize over multiple processes, note this is"
+                         "separate from threading using the proc. scheme")
```
Comment (spxiwh) [inline bin/pycbc_inspiral:201]: It's really not clear that this is the number of processes to use? You refer to "nprocesses" in help text below, but that's not what this is called. This reads more like a boolean flag from the help text.

File: bin/pycbc_inspiral [hunk 3]
```diff
@@ -371,106 +444,48 @@ with ctx:
...
+
+        for elem in data:
+            out_vals_all, tparam = elem
+            if len(out_vals_all) > 0:
+                event_mgr.new_template(tmplt=tparam)
+
+                for edata in out_vals_all:
+                    event_mgr.add_template_events(names, [edata[n] for n in names])              
+                                    
+                event_mgr.cluster_template_events("time_index", "snr", cluster_window)
```
Comment (spxiwh) [inline bin/pycbc_inspiral:473]: This is a bit hard to parse in the diff. But is it the case that `consolidate_events` will run *outside* the for loop `for elem in data` but *within* the `for tchunk in tchunks` for loop? I'm not sure `(tchunk[0]) % opt.finalize_events_template_rate` is then the right check for running consolidate events as I think there are cases where that never parses to True?

It feels like  `opt.template_groupby` and `opt.finalize_events_template_rate` might actually be combined into the one option? "Run this many templates and then consolidate" (either by stopping the multi-processing pool, and/or just by clustering triggers). I can't really see a use case where this might be set to *different* values?

Comment (spxiwh) [reply bin/pycbc_inspiral:473]: ... Although perhaps having `finalize_events_template_rate` run for every N *groups* of triggers, might also make sense. Checkpointing perhaps needs to occur less often than clustering?

File: examples/inspiral/run.sh [hunk 4]
```diff
@@ -1,16 +1,15 @@
-#!/bin/bash -e
```
Comment (spxiwh) [inline examples/inspiral/run.sh:1]: I think you've reverted a bunch of my changes to this test. It's important that we make sure this runs for different FFT threads and backends. Please incorporate changes into this if needed, but not revert (I suspect this wasn't intentional, but if it was we should discuss).

**Review by spxiwh — CHANGES_REQUESTED** [review]
Thanks for this Alex. I've left some comments and requests, but I think it's only minor stuff.

Thanks for raising the sigmasq issue, that's definitely a nasty bug, but I think one that would not have made much difference to anything. Good to have it fixed though.

It's true that this will also be present in the `skymax` inspiral code, and the sigmasq fix will probably break the `multi_inspiral` code. I think this does raise a broader question of how to handle multiple inspiral front-end codes, which all do slightly different things, but share a lot of functionality. It's messy to manually have to keep all of these up-to-date. Is there a good way to unify these codes?

File: examples/inspiral/run.sh (hunk 4 above)
Comment (ahnitz) [reply examples/inspiral/run.sh:1]: Yes, sorry, I didn't mean to commit this actually. It was just for testing beforehand. I'll revert this back.

File: bin/pycbc_inspiral (hunk 3 above)
Comment (ahnitz) [reply bin/pycbc_inspiral:473]: @spxiwh Yes, that was one of the issue I was pointing out before that if you don't set the options very specifically, I think consolidate might not run.

@spxiwh Either option you propose (i.e. combining, or having finalize act on 'groups' rather than individual templates) would be fine with me. Combining is simpler, but does it server your use case (I think @josh-willis added this function). Combining does had some restriction on the range of 'good' values, but my guess is they were in the same range already. 

 Checkpointing doesn't happen at the same rate anyway, it's just the check if it needs to do a checkpoint which does. So as long as it is not so long that it will never check, that's fine from the checkpoint perspective.

Comment (spxiwh) [reply bin/pycbc_inspiral:473]: @ahnitz I think I was the one who moved `consolidate_events` into the loop over templates, and this was done to control memory usage in particularly noisy stretches of data (it's pointless storing in memory triggers that will be clustered away later). This feels to me like it's tied to the same decision as "how many templates should I run in a parallel (or not) pool before consolidating", so I'd be happy to combine these.

Comment (ahnitz) [reply bin/pycbc_inspiral:473]: @spxiwh Ok, great. I'll combine those options then.

File: bin/pycbc_inspiral (hunk 2 above)
Comment (ahnitz) [reply bin/pycbc_inspiral:201]: I'll rename to be more verbose.

**Comment by ahnitz** [comment]
```

> Thanks for this Alex. I've left some comments and requests, but I think it's only minor stuff.
> 
> Thanks for raising the sigmasq issue, that's definitely a nasty bug, but I think one that would not have made much difference to anything. Good to have it fixed though.
> 
> It's true that this will also be present in the `skymax` inspiral code, and the sigmasq fix will probably break the `multi_inspiral` code. I think this does raise a broader question of how to handle multiple inspiral front-end codes, which all do slightly different things, but share a lot of functionality. It's messy to manually have to keep all of these up-to-date. Is there a good way to unify these codes?

I'm not sure. I think the more immediate issue is there isn't really any test of those codes that makes it easy to verify for obvious breakage. I think adding at least a simple script that can run both those executables would at least make it clear. 

Is there anything simple which you all may have internally which could serve as a test we could add? I haven't run these codes in their current state, so don't have anything immediately on hand (of course I could figure out how to run them, but if something already exists...). If so, then I think it might make sense to add that separately before this PR goes in, and then modify this PR to also address the same issue there. I assume there is no objection to addressing the sigmasq issue in those codes?

File: bin/pycbc_inspiral (hunk 1 above)
Comment (ahnitz) [reply bin/pycbc_inspiral:203]: Option removed in favor of using the finalize events template rate option

File: bin/pycbc_inspiral (hunk 3 above)
Comment (ahnitz) [reply bin/pycbc_inspiral:473]: Ok, I've updated to just use the --finalize option as this value.

**Review by spxiwh — APPROVED** [review]
This all looks good to me. Thanks.
